Screen space reflection with hierarchical ray tracer

fall2019 opened the pull request:
Also add SSR mask to choose object receiving reflection.

https://github.com/asc-community/MxEngine/assets/53587145/19f3379f-ca8f-493c-8308-5c1f32531eef

fall2019 committed:
temp

fall2019 committed:
Merge branch 'hierarchical_z_buffer' of https://github.com/asc-community/MxEngine into screen_space_ray_tracing

fall2019 committed:
max pooling

fall2019 committed:
prototype

fall2019 committed:
formatting

fall2019 committed:
add intensity

fall2019 committed:
add mask

fall2019 committed:
Delete samples/Sponza/Engine/Models/cube_2.obj

fall2019 committed:
Delete samples/Sponza/Engine/Models/plane_2.obj

fall2019 committed:
Delete samples/Sponza/Engine/Models/sphere_4.obj

fall2019 committed:
Delete samples/Sponza/sponza2.json

fall2019 committed:
edit user input and MD

fall2019 committed:
Update MxApplication.cpp

fall2019 committed:
Update MxApplication.cpp

fall2019 committed:
Update MxApplication.cpp

fall2019 committed:
Update ShaderBase.cpp

fall2019 committed:
move Camera to header file

MomoDeve commented:
Hi, thanks for PR!
Have not start in-depth review process yet, but lets do couple of things to simplify the process and speed-up the merge
1) I see there is a bunch of refactoring in shaders going on, this can be moved in a separate PR and be quickly reviewed
2) I see you introduce a constant to conditionally compile code for compute & fragment shaders. I think this can be done in a different way, create common_utils file, then compute_utils & fragment_utils and include common there. With this, compute shaders can include compute_utils, fragment - fragment_utils, and we do not need conditional computation at all

fall2019 commented:
> Hi, thanks for PR! Have not start in-depth review process yet, but lets do couple of things to simplify the process and speed-up the merge
> 
> 1. I see there is a bunch of refactoring in shaders going on, this can be moved in a separate PR and be quickly reviewed
> 2. I see you introduce a constant to conditionally compile code for compute & fragment shaders. I think this can be done in a different way, create common_utils file, then compute_utils & fragment_utils and include common there. With this, compute shaders can include compute_utils, fragment - fragment_utils, and we do not need conditional computation at all

👌I will do it.
[Refactoring of shader compilation](https://github.com/asc-community/MxEngine/pull/97)

fall2019 committed:
Refactoring of shader compilation

fall2019 commented:
> Hi, thanks for PR! Have not start in-depth review process yet, but lets do couple of things to simplify the process and speed-up the merge
> 
>     1. I see there is a bunch of refactoring in shaders going on, this can be moved in a separate PR and be quickly reviewed
> 
>     2. I see you introduce a constant to conditionally compile code for compute & fragment shaders. I think this can be done in a different way, create common_utils file, then compute_utils & fragment_utils and include common there. With this, compute shaders can include compute_utils, fragment - fragment_utils, and we do not need conditional computation at all

https://github.com/asc-community/MxEngine/pull/97

fall2019 committed:
move common include into fragment

fall2019 committed:
Merge branch 'shader_compilation_refactoring' of https://github.com/asc-community/MxEngine into screen_space_ray_tracing

fall2019 committed:
fix merge issue

fall2019 committed:
add more comments

fall2019 commented on src/Platform/OpenGL/Shaders/ssr_fragment.glsl:
```diff
@@ -1,69 +1,174 @@
+
+    float w0 = 1.0 / ho0.w;
+    float w1 = 1.0 / ho1.w;          
+    //to avoid nonlinearity,project to view space
```
Perspective division creates uneven stepping
<img width="299" alt="截屏2024-03-16 上午10 40 09" src="https://github.com/asc-community/MxEngine/assets/53587145/23375866-7fb0-4f51-b02c-36aba54c668f">

fall2019 commented:
Also added some comments to simplify the review process and reduce maintenance efforts.

fall2019 committed:
Merge branch 'master' of https://github.com/asc-community/MxEngine into screen_space_ray_tracing

fall2019 committed:
Merge branch 'master' of https://github.com/asc-community/MxEngine into screen_space_ray_tracing

fall2019 committed:
Update dirlight_fragment.glsl

fall2019 committed:
Update dirlight_fragment.glsl

fall2019 committed:
Update apply_ssr_fragment.glsl

fall2019 committed:
Update dirlight_fragment.glsl

fall2019 committed:
Update gbuffer_vertex.glsl

fall2019 committed:
Update dirlight_fragment.glsl

fall2019 committed:
Update apply_ssr_fragment.glsl

fall2019 committed:
sync master

fall2019 commented:
Conflict resolved

MomoDeve commented on src/Core/Components/Camera/CameraController.cpp:
```diff
@@ -425,11 +426,21 @@ namespace MxEngine
         return this->renderBuffers->Depth;
     }
 
+    MxVector<TextureHandle>  CameraController::GetHiZ() const
```
return by const reference, avoid copying of heap-allocated vector

MomoDeve commented on src/Core/Components/Camera/CameraController.cpp:
```diff
@@ -493,10 +508,23 @@ namespace MxEngine
+        for (int i = 0; i < this->HiZ.size(); i++) 
+        {
+            int scale = pow(2, i+1);
+            MxString tag = ToMxString(fmt::format("!camera depth lv{}", i + 1));
```
there is MxFormat which wraps fmt::format, use it instead

MomoDeve commented on src/Core/Rendering/RenderController.cpp:
```diff
@@ -352,11 +354,21 @@ namespace MxEngine
-        // generate depth texture mipmaps for post-processing algorithms. Replace later with hierarhical depth map
-        depth->GenerateMipmaps();
+        auto& shader = this->Pipeline.Environment.Shaders["HIZ"_id];
+        MxVector<TextureHandle> inputTex = { zBuffer };
```
use array here, avoid unnecessary allocations

MomoDeve commented on src/Core/Rendering/RenderController.cpp:
```diff
@@ -1044,8 +1061,18 @@ namespace MxEngine
+        { 
+            auto& it = camera.HiZ[i];
+            it->Bind(startId++);
+            shader.SetUniform(ToMxString( fmt::format("hiZTex{}", i)), it->GetBoundId());
```
MxFormat

MomoDeve commented on src/Core/Rendering/RenderController.cpp:
```diff
@@ -1626,7 +1657,7 @@ namespace MxEngine
             this->DrawObjects(camera, *this->Pipeline.Environment.Shaders["GBufferMask"_id], this->Pipeline.MaskedObjects);
 
-            this->GenerateDepthPyramid(camera.DepthTexture);
+            this->GenerateDepthPyramid(camera.DepthTexture,camera.HiZ);
```
please use auto-formatter and put spacing

MomoDeve commented on src/Core/Resources/Material.h:
```diff
@@ -59,6 +59,7 @@ namespace MxEngine
         float RoughnessFactor = 0.75f;
         float MetallicFactor = 0.0f;
 
+        bool ReceiveSSR{ false };
```
even though its totally valid, in project we usually initialize primitive types with =

MomoDeve commented on src/Platform/OpenGL/Shaders/gbuffer_fragment.glsl:
```diff
@@ -68,4 +70,5 @@ void main()
     OutAlbedo = vec4(fsin.RenderColor * albedo, emmisive / (emmisive + 1.0f));
     OutNormal = vec4(0.5f * normal + 0.5f, 1.0f);
     OutMaterial = vec4(parallaxOcclusion * occlusion, roughness, metallic, 1.0f);
+    OutSSRMask = vec4(float(receiveSSR),vec3(1.0));
```
you can directly pass float to the shader instead of bool
with bool you have additional instruction in vertex shader to convert it to float:
```
%16 = OpSelect %float %13 %float_1 %float_0
%17 = OpCompositeConstruct %v4float %16 %float_1 %float_1 %float_1
```

MomoDeve commented on src/Platform/OpenGL/Texture.cpp:
```diff
@@ -43,7 +43,7 @@ namespace MxEngine
         GL_RG16,
         GL_R16F,
-        GL_R32F,
+        GL_R32F_EXT,//cannot use GL_R32F. Why?
```
interesting, on what machine it fails?

fall2019 replied:
![image](https://github.com/asc-community/MxEngine/assets/53587145/9e369c0f-fb12-40f3-bcff-ea7a709683a3)

fall2019 committed:
use MxFormat

fall2019 committed:
getHiz() => const reference

fall2019 committed:
avoid memory alloc

fall2019 committed:
format

fall2019 committed:
fomat

fall2019 committed:
using float for bool param pass to shader

fall2019 commented:
Resolved Comments.

fall2019 committed:
add more comments

MomoDeve commented:
Hi, sorry for delaying the review!
Can you explain why you used `intensity` in CameraSSR? Its not a physical term. I suppose metalness of surface should be used instead?

fall2019 commented:
> Hi, sorry for delaying the review! Can you explain why you used `intensity` in CameraSSR? Its not a physical term. I suppose metalness of surface should be used instead?

Yeah.MetallicFactor is used in apply_ssr_fragment. That means you can set it to 1.0 to keep it physically accurate and set to other numbers when you want it more artistic.😁

fall2019 committed:
Remove intensity. Expose levels to editor.

fall2019 commented:
![image](https://github.com/asc-community/MxEngine/assets/53587145/704dba8c-2ef3-4d5a-b1c4-051136e45d0b)
Cost is way less than 1ms under level=2. 
Higher level also can produce less noisy result.
This method can be the base to do further advanced optimizations like ray reuse and prefiltered samples in the future.

MomoDeve commented:
I made a glass bricks (remove roughness texture, roughness=0, metallness=1). When I set "receieve SSR" I got degradation from 5ms to 20ms. Thats not acceptable performance. It should behave normal on all scenes

![image](https://github.com/asc-community/MxEngine/assets/40578274/1f1c47b5-b2e4-4d21-89f0-e26a1d8b82e3)
![image](https://github.com/asc-community/MxEngine/assets/40578274/bde5d915-706e-483b-891a-d2f4f3c8c9d0)

MomoDeve commented:
these artefacts under objects also need to be fixed. There should be a depth check
![image](https://github.com/asc-community/MxEngine/assets/40578274/806683f6-8093-414c-a01b-5e89d7bd969c)

MomoDeve commented:
> ![image](https://private-user-images.githubusercontent.com/53587145/316616306-704dba8c-2ef3-4d5a-b1c4-051136e45d0b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE2NTM4NDIsIm5iZiI6MTcxMTY1MzU0MiwicGF0aCI6Ii81MzU4NzE0NS8zMTY2MTYzMDYtNzA0ZGJhOGMtMmVmMy00ZDVhLWIxYzQtMDUxMTM2ZTQ1ZDBiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI4VDE5MTkwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUwMWZiMTMzMTdlY2Y3MDQxOGFlNzJkMWYxYTg3YTIyNzM1YjZkYTA1ZDUxZjc0MDFjMTQ3Mjg5MGRiYmFkM2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.7BucVxKYbypcuDBXfdKScFchKcEysaXVPopEve0miUA) Cost is way less than 1ms under level=2. Higher level also can produce less noisy result. This method can be the base to do further advanced optimizations like ray reuse and prefiltered samples in the future.

you are measuring CPU performance with that! There is no GPU performance profiling in engine yet

MomoDeve commented on src/Core/Components/Camera/CameraController.cpp:
```diff
@@ -493,10 +509,23 @@ namespace MxEngine
 
+        for (int i = 0; i < this->HiZ.size(); i++) 
+        {
+            int scale = pow(2, i+1);
```
NIT: you can do `1 << shift` to power 2s

MomoDeve commented on src/Core/Components/Camera/CameraController.cpp:
```diff
@@ -35,6 +36,7 @@
 #include "PerspectiveCamera.h"
 #include "FrustrumCamera.h"
 #include "Core/Runtime/Reflection.h"
+#include "fmt/format.h"
```
why include

MomoDeve commented on src/Core/Components/Camera/CameraController.cpp:
```diff
@@ -520,10 +549,12 @@ namespace MxEngine
         Factory<Texture>::Destroy(this->HDR);
         Factory<Texture>::Destroy(this->SwapHDR1);
         Factory<Texture>::Destroy(this->SwapHDR2);
+        for(auto& it:this->HiZ)Factory<Texture>::Destroy(it);
```
run linter! Why no spaces?

MomoDeve commented on src/Core/Components/Camera/CameraSSR.cpp:
```diff
@@ -37,31 +37,21 @@ namespace MxEngine
+    void CameraSSR::SetMaxLevel(int level) 
     {
-        this->startDistance = Max(distance, 0.0f);
+        this->maxLevel = level;
```
where are checks for level. If its set to negative?

MomoDeve commented on src/Core/Components/Camera/CameraSSR.h:
```diff
@@ -37,17 +37,16 @@ namespace MxEngine
         float thickness = 0.5f;
-        size_t steps = 10;
-        float startDistance = 2.0f;
+        float maxLevel = 2.0f;
```
why its float here, if its set as int?

MomoDeve commented on src/Platform/OpenGL/Shaders/ssr_fragment.glsl:
```diff
@@ -7,55 +7,174 @@ uniform sampler2D albedoTex;
+    case 2:return texelFetch(hiZTex1, uv / ivec2(4), 0).r;
+    case 3:return texelFetch(hiZTex2, uv / ivec2(8), 0).r;
+    case 4:return texelFetch(hiZTex3, uv / ivec2(16), 0).r;
+        //Add more levels here
```
is it valid shader? Whats returned in default case

MomoDeve commented on src/Platform/OpenGL/Shaders/ssr_fragment.glsl:
```diff
@@ -7,55 +7,174 @@ uniform sampler2D albedoTex;
                 break;
+            }
+            //stepping back 
+            TRACE_RAY(-pow(2.0, level));
```
why you do `int(-pow(2.0, level)`? Just binary shift

MomoDeve commented on src/Platform/OpenGL/Shaders/hierarchical_depth_buffer_generator.glsl:
```diff
@@ -0,0 +1,46 @@
+        maxDepth = max(maxDepth, max(row.x, row.y));
+    }
+
+    OutColor = vec4(maxDepth,0.f,0.f,1.f);
```
dont use .f, its not allowed by standard in glsl

MomoDeve commented on src/Platform/OpenGL/Texture.cpp:
```diff
@@ -205,7 +205,10 @@ namespace MxEngine
         {
         case 1:
-            dataChannels = GL_RED;
+            if (format != TextureFormat::R32F)
```
NIT: `dataChannels = (format == TextureFormat::R32F) ? GL_RED_EXT : GL_RED`

MomoDeve commented on src/Core/Resources/Material.h:
```diff
@@ -59,6 +59,7 @@ namespace MxEngine
         float RoughnessFactor = 0.75f;
         float MetallicFactor = 0.0f;
 
+        bool ReceiveSSR = false;
```
why its false by default?

fall2019 replied:
Because it's in screen space. Most objects even have no enough information on the screen. So I guess users should manually open this for floors or walls etc...

MomoDeve commented on src/Platform/OpenGL/Shaders/Library/common_utils.glsl:
```diff
@@ -3,6 +3,9 @@ struct Camera
     vec3 position;
     mat4 viewProjMatrix;
     mat4 invViewProjMatrix;
+    mat4 viewMatrix;
```
There are lots of warnings 'ignore non existing uniform' as not all shaders use these matrices.
Thats my poor design of validating shaders, but this need to be fixed in the PR.
Create a method "ignore camera uniforms" with all these parameters set for shader as ignored, and call it in each shader using camera. This will make all of them optional

fall2019 replied:
Actually we have it. https://github.com/asc-community/MxEngine/blob/599eb5f4e55c5039a04ff5fe548bfc8d43a3d58a/src/Platform/OpenGL/ShaderBase.cpp#L360 I just not notice the warnings....

MomoDeve reviewed (changes requested):
this cant be merged with poor performance

Profile shader performance on different scenes and materials!

Use third party app like nsight to measure actual gpu performance and optimize the shader

please include the measurements after fixes

fall2019 committed:
Fix artefacts

https://github.com/asc-community/MxEngine/pull/94#issuecomment-2025958522

fall2019 committed:
using bit shift

fall2019 committed:
pack depth clips into one texture

fall2019 commented:
![image](https://github.com/asc-community/MxEngine/assets/53587145/698dd4e9-9351-4187-b8f3-fa5220608242)
![image](https://github.com/asc-community/MxEngine/assets/53587145/5b902051-0574-4fa5-99ed-303e4b24d9dd)

fall2019 committed:
remove unused header

fall2019 committed:
.f => .0

fall2019 committed:
if else => ? :

fall2019 committed:
fix warnings

fall2019 commented:
Improved performance. Packed all depth maps into one texture and tested with different scenes.
Fixed artefacts under objects.
Fixed formats and warnings.
Performance issue is resolved rn. It takes from 1ms to 2ms.
![image](https://github.com/asc-community/MxEngine/assets/53587145/f4a029be-612e-4451-9c39-50be25d29453)

fall2019 committed:
Remove unused lines

fall2019 committed:
Fix format and adjust param

fall2019 committed:
change depth map packing algo

fall2019 commented:
![image](https://github.com/asc-community/MxEngine/assets/53587145/2f5175c5-f158-4989-bf7e-45c99e34d186)
